drivers/bluetooth/hci/hci_b91.c

```diff
@@ -0,0 +1,304 @@
+		BT_ERR("Not enough data for event header");
+		return NULL;
+	}
+
+	discardable = is_hci_event_discardable(data);
+
+	memcpy((void *)&hdr, data, sizeof(hdr));
+	data += sizeof(hdr);
+	remaining -= sizeof(hdr);
+
+	if (remaining != hdr.len) {
+		BT_ERR("Event payload length is not correct");
+		return NULL;
+	}
+	BT_DBG("len %u", hdr.len);
+
+	buf = bt_buf_get_evt(hdr.evt, discardable, K_NO_WAIT);
+	if (!buf) {
+		if (discardable) {
+			BT_DBG("Discardable buffer pool full, ignoring event");
+		} else {
+			BT_ERR("No available event buffers!");
+		}
+		return buf;
+	}
+
+	net_buf_add_mem(buf, &hdr, sizeof(hdr));
+
+	buf_tailroom = net_buf_tailroom(buf);
+	if (buf_tailroom < remaining) {
+		BT_ERR("Not enough space in buffer %zu/%zu",
+		       remaining, buf_tailroom);
+		net_buf_unref(buf);
+		return NULL;
+	}
+
+	net_buf_add_mem(buf, data, remaining);
+
+	return buf;
+}
+
+static struct net_buf *bt_b91_acl_recv(uint8_t *data, size_t remaining)
+{
+	struct bt_hci_acl_hdr hdr;
+	struct net_buf *buf;
+	size_t buf_tailroom;
+
+	if (remaining < sizeof(hdr)) {
+		BT_ERR("Not enough data for ACL header");
+		return NULL;
+	}
+
+	buf = bt_buf_get_rx(BT_BUF_ACL_IN, K_NO_WAIT);
+	if (buf) {
+		memcpy((void *)&hdr, data, sizeof(hdr));
+		data += sizeof(hdr);
+		remaining -= sizeof(hdr);
+
+		net_buf_add_mem(buf, &hdr, sizeof(hdr));
+	} else {
+		BT_ERR("No available ACL buffers!");
+		return NULL;
+	}
+
+	if (remaining != sys_le16_to_cpu(hdr.len)) {
+		BT_ERR("ACL payload length is not correct");
+		net_buf_unref(buf);
+		return NULL;
+	}
+
+	buf_tailroom = net_buf_tailroom(buf);
+	if (buf_tailroom < remaining) {
+		BT_ERR("Not enough space in buffer %zu/%zu",
+		       remaining, buf_tailroom);
+		net_buf_unref(buf);
+		return NULL;
+	}
+
+	BT_DBG("len %u", remaining);
+	net_buf_add_mem(buf, data, remaining);
+
+	return buf;
+}
+
+static struct net_buf *bt_b91_iso_recv(uint8_t *data, size_t remaining)
+{
+	struct bt_hci_iso_hdr hdr;
+	struct net_buf *buf;
+	size_t buf_tailroom;
+
+	if (remaining < sizeof(hdr)) {
+		BT_ERR("Not enough data for ISO header");
+		return NULL;
+	}
+
+	buf = bt_buf_get_rx(BT_BUF_ISO_IN, K_NO_WAIT);
+	if (buf) {
+		memcpy((void *)&hdr, data, sizeof(hdr));
+		data += sizeof(hdr);
+		remaining -= sizeof(hdr);
+
+		net_buf_add_mem(buf, &hdr, sizeof(hdr));
+	} else {
+		BT_ERR("No available ISO buffers!");
+		return NULL;
+	}
+
+	if (remaining != bt_iso_hdr_len(sys_le16_to_cpu(hdr.len))) {
+		BT_ERR("ISO payload length is not correct");
+		net_buf_unref(buf);
+		return NULL;
+	}
+
+	buf_tailroom = net_buf_tailroom(buf);
+	if (buf_tailroom < remaining) {
+		BT_ERR("Not enough space in buffer %zu/%zu",
+		       remaining, buf_tailroom);
+		net_buf_unref(buf);
+		return NULL;
+	}
+
+	BT_DBG("len %zu", remaining);
+	net_buf_add_mem(buf, data, remaining);
+
+	return buf;
+}
+
+static void hci_b91_host_rcv_pkt(uint8_t *data, uint16_t len)
+{
+	uint8_t pkt_indicator;
+	struct net_buf *buf = NULL;
+	size_t remaining = len;
```

Comment:
Any particular reason for this duplicate variable, when you could instead just be decrementing `len` whenever you increment `data`?